### PR TITLE
Create comparisons of normalized_mean features between different timepoints

### PR DESCRIPTION
**Pull request by alex-l-kong**
**What is the purpose of this PR?**

Addresses second part of #27. Between three sets of timepoints: primary and baseline, baseline and post_induction, and post_induction and on_nivo, the average difference across all features for each patient need to be computed and saved to a DataFrame.

**How did you implement your changes**

The `tumor_evolution.py` function contains a generalized workflow that can be modified to achieve this. The key differences added to this implementation are:

1. Use of `Patient_ID` instead of `TONIC_ID` to define the patients
2. More robust edge case handling (duplicate removal, skip over patients that don't have tissue entries in both metadata and timepoint data, etc.)
3. Explicitly iterate over the specified timepoints and compute individually, as opposed to in bulk with the unnecessary timepoints

**Remaining issues**

Not all patients contain all timepoint comparisons. Even if a patient's tissue ID exists in the metadata for a specific timepoint, it will not always appear in the timepoint data. This is why it's best to explicitly iterate over each desired timepoint pair.

Because of this, about 45 out of 123 patients don't have any valid timepoint to compare. That's >35% of all patients, and considering the small sample, it is intriguing, if not a little concerning.

### CODE REVIEW

File: python_files/supplementary_plot_helpers.py [hunk 1]
```diff
@@ -406,3 +409,106 @@ def stitch_before_after_norm(
...
+        tp_one_data (pd.Series):
+            The data for the first timepoint
+        tp_two_data (pd.Series):
+            The data for the second timepoint
+
+    Returns:
+        float:
+            The Euclidean distance between the two timepoint datapoints
+    """
+    return np.abs(tp_two_data - tp_one_data).mean()
```
Comment (srivarra) [inline python_files/supplementary_plot_helpers.py]: Isn't Euclidean distance $f(x,y) = \sqrt{(x-y)^2}$, or is this another other type of Euclidean distance?

**Review by srivarra — CHANGES_REQUESTED** [review]
Looks good, just a question about the distance metric.

File: python_files/supplementary_plot_helpers.py (hunk 1 above)
Comment (alex-l-kong) [reply python_files/supplementary_plot_helpers.py]: Good catch, this is MAE which is what was previously implemented. Updated accordingly.

**Review by srivarra — APPROVED** [review]
Looks good!

**Review by camisowers — APPROVED** [review]
Looks good to me

File: python_files/supplementary_plot_helpers.py [hunk 2]
```diff
@@ -406,3 +414,101 @@ def stitch_before_after_norm(
 
     pre_norm_tiled.save(pre_norm_stitched_path)
     post_norm_tiled.save(post_norm_stitched_path)
+
+
+def euclidean_timepoint(tp_one_data: pd.Series, tp_two_data: pd.Series) -> float:
```
Comment (ngreenwald) [inline python_files/supplementary_plot_helpers.py]: This doesn't really line up with the original implementation. Is there a reason you didn't follow that logic?

https://github.com/angelolab/TNBC_python_scripts/blob/main/python_files/tumor_evolution/tumor_evolution.py#L13

File: python_files/supplementary_plot_helpers.py [hunk 3]
```diff
@@ -19,6 +20,13 @@
     11, 12, 13, 14, 15, 17, 18, 20, 22, 23, 24, 28, 29, 30, 31, 32, 33, 34, 35,
     36, 39, 40, 41, 42, 43, 44, 45, 46, 47
 ]
+TIMEPOINT_INDICES = ["primary", "baseline", "post_induction", "on_nivo"]
```
Comment (ngreenwald) [inline python_files/supplementary_plot_helpers.py]: This should be primary_untreated, that's the main primary timepoint we've been working with

File: python_files/supplementary_plot_helpers.py [hunk 4]
```diff
@@ -19,6 +20,13 @@
...
+TIMEPOINT_PAIRS = [
```
Comment (ngreenwald) [inline python_files/supplementary_plot_helpers.py]: What was the logic for making these global variables within the helpers file?

**Review by ngreenwald — CHANGES_REQUESTED** [review]
See comments below

File: python_files/supplementary_plot_helpers.py (hunk 4 above)
Comment (alex-l-kong) [reply python_files/supplementary_plot_helpers.py]: This was mostly for convenience, before adding the `baseline to on_nivo` pair I had originally used `itertools` to compute each individual pair. Agreed this doesn't serve as much of a purpose anymore, I can leave them hard-coded inside the function instead.

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Code looks good. Instead of putting this in the supplement though, the goal is to add this to the overall pipeline so it can be used as a feature. 

Take a look through at 7_create_evolution_df.py. We'll want to add these values into the final df that is created there. Can move the code there as well.

**Comment by alex-l-kong** [comment]
@ngreenwald after giving it some thought, it will be much easier to use the original implementation as defined earlier in this PR and modify it slightly to match the workflow in `7_create_evolution_df.py`. It doesn't take long to re-merge and the implementation more closely matches what's already in `tumor_evolution/tumor_evolution.py`.

The updated `timepoint_evolution_features.csv` now contains the following (currently stored in `timepoint_evolution_features_test_aggregate.csv`):

1. A new feature for each patient and timepoint comparison called `euclidean_distance_normalized_mean_filtered_features`: the Euclidean distance between the features defined for the patient and the two timepoints in question.
2. The `raw_value` is as defined in 1. The `normalized_value` is the z-scored values across all computed `raw_values` for the `euclidean_distance_normalized_mean_filtered_features`.

The updated `timepoint_combined_features.csv` is contained in `timepoint_combined_features_test_aggregate.csv`. Once finalized, I'll update the save to overwrite the existing `timepoint_evolution_features` and `timepoint_combined_features` files.

File: python_files/generate_supplementary_plots.py [hunk 5]
```diff
@@ -36,12 +37,12 @@
 )
 
 samples_dir = "/Volumes/Shared/Noah Greenwald/TONIC_Cohort/image_data/samples"
-samples_fov = "TONIC_TMA2_R5C4"
+samples_fov = "TONIC_TMA24_R8C1"
```
Comment (ngreenwald) [inline python_files/generate_supplementary_plots.py:40]: How come this changed here?

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Code looks good

File: python_files/generate_supplementary_plots.py (hunk 5 above)
Comment (alex-l-kong) [reply python_files/generate_supplementary_plots.py:40]: This was the FOV we settled on visualizing in #17, along with the channels excluded. Somehow this didn't make it into main, so I patched it here.

**Review by ngreenwald — APPROVED** [review]

**Pull request merged**